File: javascript/6kyu/points-on-a-line.js
/*
Codewars. 24/07/20. 'Points On A Line'. 6kyu. Here we are given coordinates and our function must return true if all of them lie on a line.
For example, [[1,2], [7, 4], [22, 9]] should return true and [[1,2], [-3, -14], [22, 9]] should return false. Here is the solution I
developed to solve the challenge.
1) We need two equations in order to solve this challenge, the first is the equation for a straight line, the second is the equation to
   find the slope given two coordinates:
   1) Straight line = y = mx + b where m = slope and b = y-intercept.
   2) Slope = (y₂ - y₁) / (x₂ - x₁).
2) We define our function pointsOnLineMS, which takes an array of xy coordinates as its argument. Each sub-array holds an x coordinate as
   its first element and a y coordinate as its second element.
3) We do our first edge case check, if points is an empty array, we return true.
4) Now we create a variable called transposed where we store the points array of arrays, transposed i.e. columns turned into rows. This
   involves calling our transpose function.
5) Our transpose function takes an array as its argument and can transpose an array containing 2 columns in each row i.e. 2 elements in each
   sub-array.
6) The transposed array will contain 2 rows/sub-arrays so we initialize row1 and row2 as empty arrays.
7) We then call forEach on the array passed in (points) and for each sub-array, we push the first element (x coordinates) into row 1 and
   push the second element (y coordinates) into row 2.
8) We then create a new array using Array.from() and set its length equal to 2, we then map over the new array and make the first element
   row1 and the second element row2. This new array is our transposed array and it is returned.
9) Now points has been transposed, all x coordinates are in the first sub-array/row, and all y coordinates are in the second sub-array/row.
10) Now we handle our second edge case check, if all x coordinates in points are the same or all y coordinates in points are the same, we
    have a line, if all coordinates in all of points are the same, we have a single point.
11) We create an if block, and call the some method on transposed, we essentially uniq (remove the duplicates) from the axis sub-array by
    turning it into a set, and if the length is now 1, then all the values in that axis were the same. If this is true for any axis in
    transposed, we have a line or a single point, and thus we return true.
12) Now we've transposed points and handled edge cases, we go about finding the answer for our standard cases. We use the first part of the
    slope formula, subtracting the first x point from the second x point and the first y point from the second y point. We store the
    products in a variable called xy.
13) Now we calculate the slope by dividing the y product by the x product. JavaScript automatically gives us a float, unlike Ruby which
    performs integer division by default.
14) Now we have the slope, we find the y-intercept by taking any point's x and y coordinates and the slope, and rearranging the formula.
    We multiply the x coordinate by the slope, then subtract the product from the y value. We ceil the result and store this in a variable
    y-int.
15) Finally, we use the every method - similar to Ruby's all? method - and check whether the values we plug into our straight line
    formula are equal to y on every point, if so, points lies on a line and thus our function returns true.
16) When we use the straight line forumla in the every method, we floor the result, when we calculated the y-intercept, we ceiled the
    result, we could have also done this the other way around and got the correct result. We also could have just used Math.round on
    both to get the correct result.
*/

function pointsOnLineMS(points) {
  if (points.length === 0) {return true;}
  let transposed = transpose(points);
  if (transposed.some(axis => [...new Set(axis)].length === 1)) {return true;}
  let xy = transposed.map(axis => axis[1] - axis[0]);
  let slope = xy[1] / xy[0];
  let y_int = Math.ceil(points[0][1] - points[0][0] * slope);
  return points.every(xy => xy[1] === Math.floor(slope * xy[0] + y_int));
}

function transpose(arr) {
  let row1 = [];
  let row2 = [];
  arr.forEach((sub,i) => {row1.push(sub[0]); row2.push(sub[1]);});
  return Array.from({length:2}).map((_,i) => i == 0 ? row1 : row2);
}

File: ruby/6kyu/points-on-a-line.rb
=begin
Codewars. 23/07/20. 'Points On A Line'. 6kyu. Here we are given coordinates and our method must return true if all of them lie on a 
line. For example, [[1,2], [7, 4], [22, 9]] should return true and [[1,2], [-3, -14], [22, 9]] should return false. Here is the solution 
I developed to solve the challenge.
1) We need two equations in order to solve this challenge, the first is the equation for a straight line, the second is the equation to
   find the slope given two coordinates:
   1) Straight line = y = mx + b where m = slope and b = y-intercept.
   2) Slope = (y₂ - y₁) / (x₂ - x₁).
2) We define our method points_on_line_ms?, which takes an array of points as its argument. Each sub-array in the array holds an x and
   corresponding y point as the first and second element respectively.
3) The first thing we do is handle edge cases, returning true if points is an empty array or if all x coordinates are the same meaning
   we have a horizontal line, or if all y coordinates are the same meaning we have a vertical line. For the horizontal or vertical
   line check we transpose the array, turning the columns into rows and visa versa, this places all x points in one sub-array and all
   y points in another sub-array, then calling uniq will remove all duplicates, so if the new length of the axis is 1, we have a line.
4) Now we begin to find the slope for our standard cases. We call transpose again on points. For example, [[7, 3], [21, 9], [14, 6]]
   becomes [[7, 21, 14], [3, 9, 6]]. The first sub-array holds the x coordinates and the second holds the y coordinates.
5) Now we can find the slope by using any two points. We subtract the first x coordinate from the second x coordinate then do the same
   for y coordinates. We store the products in xy, which is an array.
6) Now to give us the slope, we simply divide the y product by the x product. We convert both values to rational numbers in order to
   perform more precise division with no rounding errors, rationals are fractions, we also could have converted both number to floats.
   The product will be a rational number also. In case our x product is a 0, which will produce a ZeroDivisionError, we use rescue to
   make the slope 0 in that case.
7) Now we have the slope, we need to find the y-intercept, the y-intercept is the position of y when x = 0. We find the y-intercept by
   using any point's y value and x value and slope with our straight line equation.
8) We use the first point, essentially doing b = y - x * m. For example, for [[7, 3], [21, 9], [14, 6]] this would be
   3 - 7 * (3/7), rounding down makes this equal to 0. So the y-intercept is 0.
9) Now we have all the ingredients we need to check that all our points are on a line. We call the all? method on points and pass in 2
   block variables, x and y, which correspond to each element in each sub-array. For each point, we check if y is equal to what is
   produced when we use the slope and y-intercept values we've found in the straight line formula y = mx + b. We round up the product of
   the forumla. If true for all points, points is on a line and thus our method returns true, if not, it returns false.
10) Note, for the y-intercept, we could have used ceil on that when rounding to produce an integer, but then we would have had to floor
    in our all? block when using our straight line formula. So we could have reversed which rounding we used. Actually, we use could have 
    just used the round method on both to get the correct result.
=end

def points_on_line_ms?(points)
  return true if points.empty? || points.transpose.any? {|axis| axis.uniq.length == 1}
  xy = points.transpose.map {|axis| axis[1] - axis[0]}
  slope = xy[1].to_r / xy[0].to_r rescue 0
  y_int = (points[0][1] - points[0][0] * slope).floor
  points.all? {|x,y| y == (slope * x + y_int).ceil}
end